### PR TITLE
Autoscaling/sizing Block previews

### PR DESCRIPTION
**Pull request by jasmussen**
This PR is a work in progress. It starts work on improving the preview component to be way more useful. GIF:

better previews

The key thing that makes this work, is to calculate what the `scale` value should be, and the way to calculate that is relatively trivial. 

In fact, the component itself can be reauthored to receive a _source width_ and a _source height_ attribute, and it will take care of itself.

For example, if you want a landscape preview, you could do this:

```
<BlockPreviewContent srcWidth={ 400 } srcHeight={ 300 } ... />
```

and we could create JS to fill in the blanks. Specifically that math is outlined here:

https://github.com/WordPress/gutenberg/compare/try/better-previews?expand=1#diff-15792e4bd87a6bedb42355c3fd500733R67

Imagine the block you're previewing is an SVG. It has to have an explicit width and height, and then it can be scaled down perfectly according to the intrinsic aspect ratio that creates. In the code shown above, we use a 4:3 landscape aspect ratio, so the block is rendered at 400x300. 

The next step is to calculate the scale value based on where the thumbnail will be shown. Right now the style variation previews are 124px wide, so the math looks like this:

`Math.min( 124 / 400 )`, which comes to `0.31`. And sure enough, when we apply `transform: scale(0.31)` the preview looks perfect:

Screenshot 2019-06-12 at 13 07 26

The only missing piece here is that the container needs some responsive magic as well. It's here: 
https://github.com/WordPress/gutenberg/compare/try/better-previews?expand=1#diff-15792e4bd87a6bedb42355c3fd500733R41

The height is 0, but a `padding-top: 75%` is applied. This padding provides the aspect ratio of the div to scale as a responsive container (see also [this for more details](http://www.mademyday.de/css-height-equals-width-with-pure-css.html)). Where does that value come from? Simple: `srcHeight / srcWidth * 100`. So for our case, it's `300 / 400 * 100` = `75`.

So for example if you wanted a portrait 9:16 preview thumbnail, you might provide these dimensions:

```
<BlockPreviewContent srcWidth={ 450 } srcHeight={ 800 } ... />
```

This would provide a padding value of 177%. And if shown as a 200px wide thumbnail, the scale factor would be 0.25.

@youknowriad does the above make sense? If you have time feel free to take a look at the PR. Otherwise I'll explore improving the component as my next step. Right now the most difficult piece for me is that the `padding-top` has to be applied to the _container_ of the preview, and the dimensions and scale factor has to be applied to the _child_ of the container. So I might end up creating additional markup inside the preview component to make this happen. 

## Terminology

The following screenshots helps to disambiguate some of the terms used to identify the different types of preview.

Annotation on 2019-08-01 at 12-58-27

## To Demo
Show Quote style before and after (big padding and large typography makes it overflow now)

## Todo

- [ ] Test @jorgefilipecosta e2e test fix and implement here
- [ ] Get more input on Unit tests and fix
- [ ] Review Riad's comment about buttons and fix as required

### CODE REVIEW

**Comment by senadir** [comment]
@jasmussen I'm working on an unreleased PR that handles some bugs in the preview scrolling or panel on general (see issue #16028), those styles will affect the overall structure of the element (a new div has to be created) also I'm switching the alignment of preview panel from absolute to using flex on the parent, I will try to test your changes with mine later this day and see how we can incorporate them both

**Comment by jasmussen** [comment]
Worth noting that next steps for this PR likely includes adding an additional wrapper regardless, because we need 1 outer wrapper for container aspect ratio padding, and an inner wrapper that holds the content and is scaled down. So that might be fine.

**Comment by jasmussen** [comment]
Alright, I've pushed a fair bit of polish to the component. The PR is now 95% complete:

better previews

What's happened:

- The CSS has been moved to the appropriate places, i.e. to the preview compoment rather than the block switcher where it was before.
- Lots of comments, lots of code cleanup.
- New variables for srcWidth and srcHeight that can be passed to the compoment — these are the dimensions at which the item to preview would be rendered at 100% size.
- A new destWidth variable, which indicates the dimensions of the size the thumbnail should be shown at. I'll get back to this in a minute.
- The scale factor necessary to scale down the thumbnail perfectly based on its source dimensions and destination dimensions.

It works pretty well:

better previews

The missing piece here is to refactor away the `destWidth` variable. It is necessary in order to calculate the transform factor necessary to scale the thumbnail correctly. Rigth now it is hardcoded, but it would be really nice if we could calculate this, since we're already in the JS doing things.

File: packages/block-editor/src/components/block-preview/index.js [hunk 1]
```diff
@@ -23,22 +23,41 @@ function BlockPreview( props ) {
...
 		</div>
 	);
 }
 
-export function BlockPreviewContent( { name, attributes, innerBlocks, settings } ) {
+export function BlockPreviewContent( { name, attributes, innerBlocks, settings, srcWidth, srcHeight, destWidth } ) {
+	// Todo: It would be nice if destWidth could be calculated here, that's the only missing piece ot making this fully responsive.
+
+	// Calculate the scale factor necessary to size down the preview thumbnail.
+	const scale = Math.min( destWidth / srcWidth );
```
Comment (jasmussen) [inline packages/block-editor/src/components/block-preview/index.js]: Note the `destWidth` here. Right now it's being passed by the page that includes the preview component. But there's no reason this has to be. Basically if we can calculate the width necessary, we can remove that variable entirely. @youknowriad or @gziolo is this something you can help me with?

Comment (getdave) [reply packages/block-editor/src/components/block-preview/index.js]: I took a quick look at this. Presumably `destWidth` is "the current width of the preview component when rendered into the DOM`? If so then we could use `offsetWidth` to get the dimensions programmatically.

I played around to see how feasible this was. Below is the result. It's my first time using hooks so it is unlikely to be usable code so I haven't pushed it, but here's a diff of something that might help. 


```diff
diff --git a/packages/block-editor/src/components/block-preview/index.js b/packages/block-editor/src/components/block-preview/index.js
index 0af013d9d..c9608dc8a 100644
--- a/packages/block-editor/src/components/block-preview/index.js
+++ b/packages/block-editor/src/components/block-preview/index.js
@@ -1,10 +1,15 @@
 /**
  * WordPress dependencies
  */
+/**
+ * External dependencies
+ */
+import React from 'react';
 import { __ } from '@wordpress/i18n';
 import { createBlock } from '@wordpress/blocks';
 import { Disabled } from '@wordpress/components';
 import { withSelect } from '@wordpress/data';
+import { useLayoutEffect, useState } from '@wordpress/element';
 
 /**
  * Internal dependencies
@@ -23,32 +28,48 @@ function BlockPreview( props ) {
 	return (
 		<div className="editor-block-preview block-editor-block-preview">
 			<div className="editor-block-preview__title block-editor-block-preview__title">{ __( 'Preview' ) }</div>
-			<BlockPreviewContent { ...props } srcWidth={ 280 } srcHeight={ 300 } destWidth={ 280 } />
+			<BlockPreviewContent { ...props } srcWidth={ 280 } srcHeight={ 300 } />
 		</div>
 	);
 }
 
-export function BlockPreviewContent( { name, attributes, innerBlocks, settings, srcWidth, srcHeight, destWidth } ) {
+export function BlockPreviewContent( { name, attributes, innerBlocks, settings, srcWidth, srcHeight } = {} ) {
 	// Todo: It would be nice if destWidth could be calculated here, that's the only missing piece ot making this fully responsive.
 
-	// Calculate the scale factor necessary to size down the preview thumbnail.
-	const scale = Math.min( destWidth / srcWidth );
-	const previewDimensions = {
-		width: srcWidth ? srcWidth : 400 + 'px', // 400x300 is provided as a 4:3 aspect ratio fallback.
-		height: srcHeight ? srcHeight : 300 + 'px',
-		transform: 'scale(' + scale + ')',
-	};
+	const previewRef = React.createRef();
+	const [ previewDimensions, setPreviewDimensions ] = useState( {
+		width: 400,
+		height: 300,
+		transform: 'scale(1)',
+	} );
 
-	// We use a top-padding to create a responsively sized element with the same aspect ratio as the preview.
-	// The preview is then absolutely positioned on top of this, creating a visual unit.
-	const aspectPadding = Math.round( srcHeight / srcWidth * 100 );
-	const previewAspect = {
-		paddingTop: aspectPadding + '%',
-	};
+	const [ previewAspect, setPreviewAspect ] = useState( {
+		paddingTop: '75%',
+	} );
+
+	useLayoutEffect( () => {
+		const destWidth = previewRef.current.offsetWidth;
+
+		// Calculate the scale factor necessary to size down the preview thumbnail.
+		const scale = Math.min( destWidth / srcWidth ) || 1;
+
+		setPreviewDimensions( {
+			width: srcWidth ? srcWidth : 400 + 'px', // 400x300 is provided as a 4:3 aspect ratio fallback.
+			height: srcHeight ? srcHeight : 300 + 'px',
+			transform: 'scale(' + scale + ')',
+		} );
+
+		// We use a top-padding to create a responsively sized element with the same aspect ratio as the preview.
+		// The preview is then absolutely positioned on top of this, creating a visual unit.
+		const aspectPadding = Math.round( srcHeight / srcWidth * 100 );
+		setPreviewAspect( {
+			paddingTop: aspectPadding + '%',
+		} );
+	}, [] );
 
 	const block = createBlock( name, attributes, innerBlocks );
 	return (
-		<div style={ previewAspect } className="editor-block-preview__container" aria-hidden>
+		<div ref={ previewRef } style={ previewAspect } className="editor-block-preview__container" aria-hidden>
 			<Disabled style={ previewDimensions } className="editor-block-preview__content block-editor-block-preview__content editor-styles-wrapper">
 				<BlockEditorProvider
 					value={ [ block ] }
diff --git a/packages/block-editor/src/components/block-styles/index.js b/packages/block-editor/src/components/block-styles/index.js
index a1289c27c..10ef682b3 100644
--- a/packages/block-editor/src/components/block-styles/index.js
+++ b/packages/block-editor/src/components/block-styles/index.js
@@ -133,7 +133,7 @@ function BlockStyles( {
 								innerBlocks={ block.innerBlocks }
 								srcWidth={ 400 }
 								srcHeight={ 300 }
-								destWidth={ 124 }
+
 							/>
 						</div>
 						<div className="editor-block-styles__item-label block-editor-block-styles__item-label">

```

File: packages/block-editor/src/components/block-preview/index.js [hunk 2]
```diff
@@ -23,22 +23,41 @@ function BlockPreview( props ) {
...
+	const previewDimensions = {
+		width: srcWidth ? srcWidth : 400 + 'px', // 400x300 is provided as a 4:3 aspect ratio fallback.
```
[the remark this answers is not in the document]
Comment (jasmussen) [reply packages/block-editor/src/components/block-preview/index.js]: Thanks a bunch for the review! 

> Presumably destWidth is "the current width of the preview component when rendered into the DOM"

Correct, the scale factor which is key to ensuring a "perfect" preview, is calculated based on the source/render width (pretend the thing we want to preview is an image, and give it a width and height), and the destination width. The difference between those two gets us the scale factor.

Thanks so much for the diff — I'll give it a spin!

**Comment by jasmussen** [comment]
I tried applying your patch, but must've mangled it somehow, so I wasn't able to verify whether it worked as it should. Would definitely appreciate some code help. @gziolo my favorite :D do you have time for some feedback there?

File: packages/block-editor/src/components/block-preview/index.js (hunk 1 above)
Comment (jasmussen) [reply packages/block-editor/src/components/block-preview/index.js]: You know, feel free to push your changes! If it gets us closer, that's great, if not we can always revert! It's still a draft PR after all :)

**Comment by mtias** [comment]
Thanks for looking at these. They are a central piece for the experience.

**Comment by jasmussen** [comment]
Alright friends, I think this is ready for review!

The code that was added by @getdave (thank you and props) works great. I added a few tweaks, some cleanup, and a stub README. As it stands, this now works well. 

Remaining tasks are:

- Code review
- Figure out why the tests fail
- I could use help with the README

**Comment by jasmussen** [comment]
Here's a little GIF as a tease:

preview

File: packages/block-editor/src/components/block-preview/index.js [hunk 3]
```diff
@@ -23,22 +28,58 @@ function BlockPreview( props ) {
...
+	} );
+
+	useLayoutEffect( () => {
+		const destWidth = previewRef.current.offsetWidth;
+
+		// Calculate the scale factor necessary to size down the preview thumbnail.
+		const scale = Math.min( destWidth / srcWidth ) || 1;
+
+		setPreviewDimensions( {
+			width: srcWidth ? srcWidth : 400 + 'px', // 400x300 is provided as a 4:3 aspect ratio fallback.
```
Comment (mtias) [inline packages/block-editor/src/components/block-preview/index.js]: Could these be specified in the function argument as defaults? The number is also repeated in the fallback dimensions.

File: packages/block-editor/src/components/block-preview/index.js [hunk 4]
```diff
@@ -23,22 +28,58 @@ function BlockPreview( props ) {
...
-		<Disabled className="editor-block-preview__content block-editor-block-preview__content editor-styles-wrapper" aria-hidden>
-			<BlockEditorProvider
-				value={ [ block ] }
-				settings={ settings }
-			>
-				<BlockList />
-			</BlockEditorProvider>
-		</Disabled>
+		<div ref={ previewRef } style={ previewAspect } className="editor-block-preview__container" aria-hidden>
+			<Disabled style={ previewDimensions } className="editor-block-preview__content block-editor-block-preview__content editor-styles-wrapper">
```
Comment (mtias) [inline packages/block-editor/src/components/block-preview/index.js]: This is not being introduced now, but these classes are all over the place namespace wise.

File: packages/block-editor/src/components/block-preview/index.js [hunk 5]
```diff
@@ -1,10 +1,15 @@
 /**
  * WordPress dependencies
  */
+/**
+ * External dependencies
+ */
+import React from 'react';
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-preview/index.js]: Why importing `react`? `createRef` is available in `@wordpress/element`

File: packages/block-editor/src/components/block-preview/index.js [hunk 6]
```diff
@@ -23,22 +28,58 @@ function BlockPreview( props ) {
...
+			transform: 'scale(' + scale + ')',
+		} );
+
+		// We use a top-padding to create a responsively sized element with the same aspect ratio as the preview.
+		// The preview is then absolutely positioned on top of this, creating a visual unit.
+		const aspectPadding = Math.round( srcHeight / srcWidth * 100 );
+		setPreviewAspect( {
+			paddingTop: aspectPadding + '%',
+		} );
+	}, [] );
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-preview/index.js]: Is this only supposed to run once, what if `srcWidth` and `srcHeight` change? should we add them to the dependencies array here?

File: packages/block-editor/src/components/block-preview/index.js (hunk 5 above)
Comment (getdave) [reply packages/block-editor/src/components/block-preview/index.js]: Code review ftw. Thanks 👍

**Comment by getdave** [comment]
@jasmussen I'm struggling with how best to manually test this in browser. Might sound silly but do you have a nice workflow to test?

**Comment by jasmussen** [comment]
For now I've simply just verified that every piece of the editor that uses this component looks right. That is, open the block library and hover over a reusable block, or open the switcher or styles panel for a block that has variations, such as Quote or Button.

**Comment by mtias** [comment]
Can we address the pending comments and get this in?

**Comment by getdave** [comment]
> Can we address the pending comments and get this in?

I will look at these today 🤞

File: packages/block-editor/src/components/block-preview/index.js (hunk 3 above)
Comment (getdave) [reply packages/block-editor/src/components/block-preview/index.js]: Fixed

File: packages/block-editor/src/components/block-preview/index.js (hunk 4 above)
Comment (getdave) [reply packages/block-editor/src/components/block-preview/index.js]: OK I've updated `editor-` to `block-editor-` to reflect the package.

File: packages/block-editor/src/components/block-preview/index.js (hunk 6 above)
Comment (getdave) [reply packages/block-editor/src/components/block-preview/index.js]: Nope. Great spot @youknowriad. Told you my hooks work needed a review! 

Fixed.

File: packages/block-editor/src/components/block-preview/index.js (hunk 5 above)
Comment (getdave) [reply packages/block-editor/src/components/block-preview/index.js]: Fixed

**Comment by getdave** [comment]
Unexplained test failures similar to issues uncovered at

https://github.com/WordPress/gutenberg/pull/15485

@jorgefilipecosta has agreed to take a quick look later this week.

**Comment by senadir** [comment]
@getdave how should I reimplement #16697 since they mostly fix the same things + other outstanding issues 
as for the failing test they could be because of the inconsistency of e2e tests, try running them again, since the same code passed for me

**Comment by youknowriad** [comment]
Capture d’écran 2019-07-24 à 11 55 41 AM

Is this normal? The buttons are too small?

**Comment by getdave** [comment]
> @getdave how should I reimplement #16697 since they mostly fix the same things + other outstanding issues

Pull down my PR. Implement your changes and push them back up. Ideally, coordinate with me here via comments to ensure we don't clash.

Would be great to keep these improvements in a single PR.

**Comment by getdave** [comment]
> Is this normal? The buttons are too small?

@youknowriad I don't _think_ it's normal. Will need to run some tests. I guess this is why we need some more design feedback.

**Comment by senadir** [comment]
@getdave how about the changes that handle the popover components & fix the scrolling problems, they're not related directly to the preview, should I push them here?

**Comment by getdave** [comment]
> they're not related directly to the preview, should I push them here?

@senadir I'd say don't push them here. If your PR merges I'll rebase.

**Comment by getdave** [comment]
I'll be picking this up again this week.

**Comment by getdave** [comment]
@shaunandrews 👋 I could really do with some designer eyes on this to help me test and review it. 

Main points:

* Are the previews rendering as we expect (see @jasmussen's description above)
* Any edge cases that aren't working?
* Is the documented API clear enough?

Much appreciated.

**Comment by getdave** [comment]
> Is this normal? The buttons are too small?

@youknowriad Following testing, I can confirm there appears to be an issue whereby if the Block being previewed has is very small it looks odd. For example try adding the button block and the opening the Block Styles Switcher. You will see it is fairly tiny which looks bad.

Screen Shot 2019-07-30 at 14 50 37

Also happens with the quote Block:

Screen Shot 2019-07-30 at 14 49 51

Maybe we need to enforce a minimum size?

**Comment by shaunandrews** [comment]
image

I noticed when adding a Separator block, the transform/style popover is broken in this branch.

**Comment by senadir** [comment]
@shaunandrews it's a problem with overflows, It's fixed in @16697 but due to how many things required to be structurally changed, it conflicted with this PR (even though I based it upon this one)
I need to do some cleaning first, separate the code from this branch & that branch, & try to do a new PR and merge again

**Comment by shaunandrews** [comment]
Like others above, I noticed that some smaller blocks are shown very tiny. And since the container has rounded borders, things like the squared button style aren't shown right in the small thumbnail.

I wonder if it'd be possible to center the blocks within the container for both the smaller thumbnails and larger previews. Also, for smaller blocks (like button and separator) can we should the larger preview at full size?

I also question the usefulness of the border on the flyout preview.

preview

**Comment by obenland** [comment]
> Like others above, I noticed that some smaller blocks are shown very tiny.

Yeah that's due to the default size being 400px by 300px. I wonder if we could make a more educated guess to the actual size of the block (rather than the room it has).

**Comment by getdave** [comment]
> I wonder if it'd be possible to center the blocks within the container for both the smaller thumbnails and larger previews. 

@shaunandrews I can take a look at that.

> Also, for smaller blocks (like button and separator) can we show the larger preview at full size?

I'm working on an implementation which I pushed in 

https://github.com/WordPress/gutenberg/pull/16113/commits/4939fd9931e4b0eb3a4d07621aff3ec28ae0a56c

It's still WIP, but it actually gets the value of the Block contents (rather than the surrounding divs) and uses that to determine the scale factor. This means that it shows things like Button zoomed in but for other types of Block it remains zoomed out as normal.

This is still a WIP and is very flaky. I'm looking to @youknowriad for assistance with this.

**Comment by getdave** [comment]
@youknowriad Can you advise as to why I need the `setTimeout()` in the `useLayoutEffect` to avoid `blocksRef.current` being `null`?

https://github.com/WordPress/gutenberg/pull/16113/commits/4939fd9931e4b0eb3a4d07621aff3ec28ae0a56c#diff-acdf0d3d62ecf51cf07d010191021618R87

**Comment by youknowriad** [comment]
@getdave I think that's because `ScaledPreview` is not rendered directly but there's some behavior in `BlockEditorProvider` only rendering it after the editor is "ready".

The way to fix this is to move the hooks, refs... to `ScaledPreview` directly.

**Comment by getdave** [comment]
The original aim of this PR was to scale the preview of a Block down so that it always fitted within the element which contained the preview.

Currently on `master` previewing larger Blocks or those with a lot of content causes the preview to bleed outside of the edge of the visible preview area (ie: you couldn't see some of the preview).

This was/has been solved by dynamically calculating a `transform: scale()` to scale down the preview component to fit within the dimensions of it's containing element.

However, there is now a problem.

Single Blocks whose contents don't occupy much visual width (eg: Button and Separator Blocks) now look a little...odd.

Taking Button as an example: creating a Button Block results in the preview (correctly) scaling in such a way that the visual representation of the button is in the top left of the preview and it is (relatively) quite small.

Whilst this is correct, to the user this may appear odd.

An attempt to handle this has been started, whereby we reach into the DOM and attempt to "measure" the element which best represents the _visual_ content of the Block. For the button Block this element is:

`div.editor-rich-text.block-editor-rich-text` 

By measuring the computed size of this element we can use this to determine the scale factor such that the preview is zoomed in on the _visual part of the Block_ (ie: the bit that actually looks like a button!), rather than the Block + the Block editing "chrome" (which is often much wider than the visual part, but is "invisible").

The major problem with this is that there is no consistent way to target the element that represents the visual part of the block. So whilst for the Button block it is the element corresponding to the selector `div.editor-rich-text.block-editor-rich-text`, for the Separator Block it is sometimes a `:before` pseudo element and sometimes a `<hr>` element.

This lack of consistency makes it impossible to write catch all logic which will provide a faithful "zoom" for all Blocks.

If we can remove the inconsistency and provide some kind of hook that would allow us to accurately target the element that represents the rendered visual size of a Block contents then we could expland on this approach. If not then I cannot see how this approach will ever work.

Therefore, I'm proposing that the only solution is to blacklist certain Blocks from being dynamically scaled _at all_. For such Blocks we would retain the existing _hardcoded_ `scale` factor. We would also only apply this logic if there is only 1 Block being previewed.

Matias has suggested using ShadowDom and iFrames but I'm not clear how this will help in this scenario. I could well be missing something here so I'd value any insight and direction if these provide a solutoin to the problem I've outlined above.

We now need to agree a direction in order to move this forward.

**Comment by obenland** [comment]
> Matias has suggested using ShadowDom and iFrames but I'm not clear how this will help in this scenario. I could well be missing something here so I'd value any insight and direction if these provide a solutoin to the problem I've outlined above.

@mtias Could you share your thoughts on that in a little bit more detail? How could ShadowDom and iFrames be used to solve this?

**Comment by jasmussen** [comment]
Mostly AFK this week, back next, so the following is a brief take that may lack context, so apologies for that. 

The preview component has two aspects to it. One is the source dimensions, one is the target dimensions, and the scale factor is calculated based on that.

Think of the source dimensions as the intrinsic width and height of an image. Say you have a 400x300 PNG image, you size that down to 100x75px at a .4 scale factor.

I believe we currently hardcode those source dimensions to a "goldilocks" value, I can't recall the precise dimensions. But those source dimensions can be manipulated too. So for example a button could be rendered with source dimensions of 200x100, which means it would fill up much more of a scaled down version. 

How we supply those different source dimensions, is a tricky one. A template preview definitely needs big source dimensions, say 800x1200, whereas as mentioned a button needs a small one. Could be autodetected? Could be a block prop?

**Comment by getdave** [comment]
Thanks @jasmussen.

> How we supply those different source dimensions, is a tricky one.

This is _precisely_ the problem. You're summary is correct and you are correct that being able to manipulate these source dimensions separately for each Block is the solution. The problem is there is no way to specify this dynamically. 

[See here](https://github.com/WordPress/gutenberg/blob/cd3baf4d9999886b7b7a3add730456d56b75c373/packages/block-editor/src/components/block-styles/index.js#L127) for how the previews are rendered for the Styles Preview. How could we specify different `src` values here on a per Block basis?

> Could be autodetected?

I've spent all day trying to do this consistently. I hope the comment above explains why this can't easily be done (although it's quite difficult to explain!).

> Could be a block prop?

This would work, but I'm doubtful we'll want to increase the API surface area.

**Comment by getdave** [comment]
I've been thinking some more about this and I have a new proposal.

If we leave the small thumbnail previews "as is" with a hardcoded `scale` factor (as per `master` branch) and only scale the large `Preview` popover?

That way we retain the existing zoomed in experience whilst also providing a nicely scaled full preview. 

This makes a lot of sense when you consider that there is little point in having both thumbnail and preview look the same. One should provide an alternative presentation / view of the preview.

If we enable this via a prop (eg: `shouldScale: true`) then we can allow scaling to be selectively enabled/disabled depending on context. 

Here are some screen shots of how this could look

Screen Shot 2019-07-31 at 16 57 31
Screen Shot 2019-07-31 at 16 54 56

Note how only the large popover preview (on the RHS) is dynamically scaled to fit.

I'd greatly appreciate some feedback on this approach.

**Comment by jasmussen** [comment]
IMO that seems like a good first step to unblock this improvement for landing, and it doesn't preclude further explorations in the future.

**Comment by retrofox** [comment]
> ...using ShadowDom and iFrames but I'm not clear how this will help in this scenario.

I've added [a commit which renders the `<ScaledPreview />` into a shadow dom element](b7fd4e7bc4e285a5da2847d3f96ecef037fbdb00).
I guess the benefits of using this is to unlink the styles between the primary one and the styles which we could apply to the shadow dom. At the moment I haven't added any CSS to this one so the preview shows an unstyled component at all.

image

**Comment by marekhrabe** [comment]
I'm not sure whether it was intentional or not but I've just noticed that this PR removes a way to use unscaled preview. Previously, the scaled preview (statically set to be 90% of original size) was used for thumbnails in block inserter/switcher and the unscaled preview was used in the window that shows when you hover reusable block/style. 

This PR seems to always need `srcWidth` and `srcHeight` but I think we should allow for usage without any scaling.

**Comment by marekhrabe** [comment]
Trying to take a step back, the biggest problem seems to be that small things look way too tiny in thumbnails. Poor button block here:



Those thumbnails currently use 400x300 as their source size. I have tried using 200x150 src size in `block-styles/index.js` and results are:

Button with small label:

Screenshot 2019-07-31 at 15 50 37

Button with huge label: 

Screenshot 2019-07-31 at 15 51 20

Divider:

Screenshot 2019-07-31 at 15 52 14

Short quote:

Screenshot 2019-07-31 at 15 52 25

Long quote:

Screenshot 2019-07-31 at 15 52 32

Nowhere near perfection but I think these are already a tad better than what we have in master (where the large quote won't even be visible).

What if we just pick less dramatic scaling factors and try to roll with it?

**Comment by getdave** [comment]
> I've added a commit which renders the <ScaledPreview /> into a shadow dom element.

@retrofox Ok - thanks for experimenting here. Is this based on any further information from @mtias as to why that would be a good idea?\

> I guess the benefits of using this is to unlink the styles between the primary one and the styles which we could apply to the shadow dom.

I'm unclear as to why that would be desirable. We want to display the styles that come with the Block. This becomes especially important when you consider we cannot know in advance the styles of every Block, particularly 3rd party ones.

The issue as I understand/see it isn't one of isolating styles but rather targeting DOM elements. It is possible I've missed something however...

> I'm not sure whether it was intentional or not but I've just noticed that this PR removes a way to use unscaled preview. 

@marekhrabe I added a prop to allow toggling the scaling on/off. This is exactly what you're proposing. If it's been removed I'll reinstate it, largely because of the reasons you highlight.

> What if we just pick less dramatic scaling factors and try to roll with it? What do you think? We can also tweak the scale factor separately for the large preview that shows up once you over things.

This is basically what I suggested in [this comment](https://github.com/WordPress/gutenberg/pull/16113#issuecomment-516917681), but with an additional improvement of tweaking the ratios which I really like. 

Due to lack of any other feasible direction and indicative support/approval from @jasmussen, I'm going to roll with this approach today.

**Comment by retrofox** [comment]
> @retrofox Ok - thanks for experimenting here. Is this based on any further information from @mtias as to why that would be a good idea?\

No, any extra information from Matias. What I did were learning and experimenting with shadow DOM for a short time since I hadn't clue how it works. After reading and playing with this, conclude that although it could be a very powerful resource to handle components previsualization it's clearly out of the scope of this approach.

> I'm unclear as to why that would be desirable. We want to display the styles that come with the Block. 

Being able to apply global and specific styles in a solid way to components.

The idea behind using this approach is selecting specific styles to visualize each component. Generally, need to use global styles for all components (current one), but sometimes it could be useful to isolate specific styles for each component. Shadow DOM handles this case in an extremely safe way. 

Sorry by my mistake to not being 100% clear about experimenting around the shadow dom. Next time I'll create another branch.

**Comment by obenland** [comment]
Seems like there is a unit test failing currently:

```
FAIL packages/block-editor/src/components/block-preview/test/index.js (14.755s)
  ● BlockPreviewContent › renders a preview with suitable default dimensions

    Method “props” is meant to be run on 1 node. 0 found instead.
      26 | wrapper.update();
      27 |
    > 28 | const previewAspectRatioStyle = wrapper.find( '.editor-block-preview__container' ).first().prop( 'style' );
         | 		                                                                                           ^
      29 | const previewTransform = wrapper.find( '.editor-block-preview__content' ).first().prop( 'style' ).transform;
      30 | const previewScale = parseInt( previewTransform.match( /\d/ )[ 0 ], 10 );
      31 | 
```

**Comment by getdave** [comment]
The direction of this has pivotted and is now running over at https://github.com/WordPress/gutenberg/pull/16873

**Comment by obenland** [comment]
Closing in favor of #16873.
